### DISCUSSION TITLE
why the difference of getting a neode between C++ version(through browse path) and Python version(through Node Id)

### DISCUSSION THREAD

**Discussion by rockonedege** [FreeOpcUa/opcua-asyncio#1031] [Q&A]
I notice the different way to get a node. I understood the c++ version is not maintained any more.

However, I am curious whether this is a carefully thought out choice, e.g. out of performance, OPC UA standard compliance, ease of development etc., or just a random pick.  

This is to learn from design decisions.  
  
- In Python
```python
    def get_node(self, nodeid: Union[ua.NodeId, str]) -> Node:
        """
        Get node using NodeId object or a string representing a NodeId.
        """
```
e.g.

```python
client.get_node('ns=4;i=1013')
```

- In C++
```c++
  //The GetChildNode methods return a node defined by its path from the node. A path is defined by
  // a sequence of browse name(QualifiedName). A browse name is either defined through a qualifiedname object
  // or a string of format namespace:browsename. If a namespace is not specified it is assumed to be
  //the same as the parent
  Node GetChild(const std::vector<OpcUa::QualifiedName> & path) const;
  Node GetChild(const std::vector<std::string> & path) const;
  Node GetChild(const std::string & browsename) const ;
```
e.g.
```C++
    auto c_semaphore = handshake.GetChild({"Header", "Command", "Semaphore"});
```

**Comment by schroeder-** [comment]
In this library you can also get nodes via browse path:
`node = await client.nodes.root.get_child(["0:Objects", f"{idx}:MyObject", f"{idx}:MyVariable"])`
The server always use the nodeid to access any info. So calling **get_child** is a call to the server to translate it to the nodeid. So this cost a network roundtrip. The service behind that is called [TranslateBrowsePathsToNodeIds]( https://reference.opcfoundation.org/v104/Core/docs/Part4/5.8.4/). 
If you use BrowsePath or NodeId depends on your preference and sometimes on the addresspace layout of the server. So if you know the nodeid before, its best to use the nodeid. If the server addresspace use random nodeids but defined BrowseNames, this is the case in a lot of companion specifications, then use BrowseNames.

**Reply by rockonedege** [reply]
This explanation matches my understanding that opc-ua address space is a graph(rather than tree) of nodes connected with references.  Then why the C++ version interface prefers browser paths, even though it's less efficient and it could print out node id(clearly it knows the id to a node)

* graph, tree in the sense of computer science

**Reply by schroeder-** [reply]
Thats correct it's a graph. As I sad it is more a preference. But in some cases it's usefull because you may know the BrowseName but not the NodeId. For example you have a Object and the NodeId, so you can access the know child property via BrowseName instead of getting all references via get_children. 
Another reason can be that it matches the tree in ua expert, so for tutorials its easier to follow.
